pipeline/src/main/java/edu/illinois/cs/cogcomp/pipeline/server/MainServer.java

```diff
@@ -77,23 +96,58 @@ public static void startServer(String[] args, Logger logger) {
 
         port(parseResults.getInt("port"));
 
+        // create a hashmap to keep track of client ip addresses and their
+        int rate = parseResults.getInt("rate");
+        if( rate > 0) {
+            clients = new HashMap<String, Integer>();
```

Comment:
If SparkJava handles requests concurrently, you should use a ConcurrentHashMap here.